packages/@react-stately/table/src/useTableState.ts

```diff
@@ -53,19 +55,23 @@ const OPPOSITE_SORT_DIRECTION = {
  * Provides state management for a table component. Handles building a collection
  * of columns and rows from props. In addition, it tracks row selection and manages sort order changes.
  */
-export function useTableState<T extends object>(props: TableStateProps<T>): TableState<T> {
+export function useTableState<T extends object, C extends TableCollection<T> = TableCollection<T>>(props: TableStateProps<T, C>): TableState<T> {
   let [isKeyboardNavigationDisabled, setKeyboardNavigationDisabled] = useState(false);
-  let {selectionMode = 'none'} = props;
+  let {selectionMode = 'none', collectionFactory} = props;
 
   let context = useMemo(() => ({
     showSelectionCheckboxes: props.showSelectionCheckboxes && selectionMode !== 'none',
     selectionMode,
     columns: []
   }), [props.children, props.showSelectionCheckboxes, selectionMode]);
 
+  let defaultFactory = useMemo(() => (
+    (nodes, prev) => new TableCollection(nodes, prev, context)
+  ) as CollectionFactory<T, TableCollection<T>>, [context]);
+
   let collection = useCollection<T, TableCollection<T>>(
     props,
-    (nodes, prev) => new TableCollection(nodes, prev, context),
+    collectionFactory?.(context) || defaultFactory,
     context
   );
```

Comment:
To make this slightly more extensible still, I'd like to make the prop `collection` rather than `collectionFactory`. This would allow collections to be generated in other ways than only our JSX-based CollectionBuilder. If you did want to use our JSX-collection builder, you'd call `useCollection` yourself, and pass the result in as the `collection` prop.

To do this, we'll have to make `useCollection` accept `props.collection` and return early in its `useMemo`. Let me know if this doesn't make sense.

Reply:
@devongovett, Well, this may by a point, but I see several issues with this solution
- If we look at `useCollection` as independent hook, `collection` property just literally makes it do nothing, which could be confusing.
- At this moment `useCollection` requires `children`, in case when `collection` is passed this requirement looses it sense. It probably could be fixed with help of discriminating unions, but it will lead to rewriting lots of types, cause now they made as `interfaces` and unions are `types` and interface can not extend unions. It looks like a huge change for such tiny feature.
- In case we reduce types accuracy and just make `collection` optional property, It also leads to types problems, cause explicit generic declaration for `useCollection` required ([playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwNIFMCeAaOAlFBDAYxgDkIATFLXQmAUQBsUQUA7GLAVwGcVcAzTjwCyTCAF84fKBBBwA5FHxE5AbgBQa4GxRQ+hFHFIUAPABUAfIjVw4AegBU9uKYAWBmGjAGIfOMBhMcDAuwFxwLOQGimCKPGxcAHRw9rbWQZ4oAFxwXDBQWgDmGGkOTgCCcBwswACOHAYA1uiS0EFu4ZFJKWlNaNmomCWOzu0QAEYAVihEcABuePT1bQYRFHAA7nhhBIp4AWSS0iBdqTbzi1nOxTalIwaMsyj0cD5wFGDBbaEda997fixlnA3MAoHgoAQXGgTmkHk9siwOCAxjprnZhgB1NzBHRfML+QIuLZwSHAehkRQsLAoR6A4C+CLwegQPAUA5oFAwGE2IlcADCIXJRhQXGyYwgEEYeCpQycrnuLLZJMFFNYL18wW+qxQ3OVZLIwtFcAAkgEwWNGMZhWZzOY0bd5XBKRRFAcCBBtPF1XifgYABQ6gpJABSAGUABoASl1zp0KDI2WoRGF9uGFVy+RYBTmCyWfFamrC2sE8ZaUEk+BgHFicHowCa6S8eDcrN1AQAHjAAGq5y4Zwqp8qAwgEEVcYBjMn+NC1vAo5758uF326uTg4B4AC09DnTzkAH5sv2s4O7gCKO3vcvtRt-CE6TAwmBwawuclTueUO3D+EkSioKeFR8FURDAB6bT-FwLgQBw5JwCiJILIwBwwBAGxgmAQKxq6K7vmk6wYV4UA-gGjDMGwiZKHQZGvpGcAALyWEm1FMK+p6Or0V7tM+lLwNquo8a+Aw-gM7HtJxrw4r68EoIu7ghEWnR4TYMQ0sJ-ToGJjTNJJ7Q3ngfBmj6-HKeEX4wOpcCibKcBlGQZD+GBLALHAMQQERMDACKOReAQ9LAAQQRoXgrngp5BCweC0keF4AnSGAXA-tKgw3MMAACMTAPMASmVBMHksaLDnMAZB7CgJHutonbZFUDQROsLB0Yx8ESlKLBqGIGhaGaeijnAfKSowoEejacDmawZBhKaOhzpaFhWKlcp6X+uKvASID4oCUnuvQQ2eR6Mb4GQHr0DO44AF6XIiyKohoi0mmaZUvI85ZIXAvSbUCO17U5uoFJyAxcH6kbZNN5qWgMdp3eiTgAOKcnA0p+AEshjDO-hhL0f2ctNIB+r0GloCDVzQ7c8PwEj63wTOUlaBe3oY+96C6nsfole2CIrVAxOmMUNnk0CnHBP87rMGEKJyUCBRZWqnFaF9g3TPtLDYxZ6AAEKydAKD4+ghPEwMcAAD6-rtfP3QLUlC0S8Ci95BlGVJ0u0kz6Nbe031K79pn-WraBlIZOi630VnoAbzQm4iZuk8MlvtHwoK5K7AIK7tXsHT7nIAGKJ37wOE8bpv0ObMNwHH9xbPAcvuwYnvDSrmcwAAMpXAz56HM6R7B9AdV12i6PoD1MAACvFXCjQgNm4CwLqlpVARerpBhU60pJwetiSmWvqosJRNAplPE1xm6HoL4+XHLyjfi+AABtvlI3ySp94Fo2wqsjTCb90NieTAjA-sxFMdhbCznnPufcT9PSPn3DZdMeRChOhQKpOIMA9hOQvh-EAchtin1fFwEsBxFwVj2NWbydYGwxXwC2MgbZzI9guD+Y8RRYHDgIKOLg45Jx1g8KAp4ZYfTrVXOuLcO55wHiPPAk8sDayhHgK8bemCXiTCVpvAAqjwQhrQyBoGciAAKT8071y-h+beuNEqg0ehaFANpTxYk5G4JcClFG8j1OSSk1IXb0g6EyRUpYORvm-sCLYAp9RmJ-OKSU+B2qdTUJQoeIAGCsTYKNeiOAqKJPIjAYwuNR7uXHhYcw6g4m42nrPKAKS4BsxRtkUwTVLC4wya+G06hND916gYUM6cWC5IShPQ+M9j6QLPmEJeOQunUgSEGRGwIjo6DbP4f+e9kyREAqw9hnCpw8NEXwohUkeD1yEfkERu56DiJyJI5h91QyoPCq4g4G8XivUkCBZWQVEEDPLPfVguovm73iY05J80TYNJooC8wABtAAuoXEpR9FDlIsKeXGSiphEE+nsrpuoN4-jBrNaxFhe6xIyHATp9cAVZPmqk5i5LjCkuVj0-JtoWlxIGoY5WNLKUkq6RyywwKUY8qKcS1lP0PQhLcawCpwruWgopby-qisyUyptFCwufoqlMBqXU+VbKnI8sjC07qOh2napFSwdWWx8VYD5GNTsE0whSvrlaSINpbQLVLo6eeeCMF12Vj8lUlJsgOuVmKneNibLIvGKi8+8ttoKt9aZbFFiZpWLDfdR0VMPoQQpooN4oQ8U0LuDwTB2xpSMhkmMva8YsD5gihorArQIA4igOsUIy9+40FLC24IuoHJcHzYDHFliIboChvdUeKAxgcH1CSbgqFZA+u9oEhdHofx8kJSyuNTks40GgGgMwVqbUL0miarpTqTAFMsKkv02ok3g2sdaApWBVKzEDYXKO9AtV8haRQAgO4c3ARYPXSoPAg1OTMIeu1KKlYHvGjPe1m6RoPtdak0DiHnUXr9GkFSY9A0IbNRa-d-U7RwCw5IHdUA0Crrw9uogu7CN8mI6Rz1nYfy1XqlSEjNgbBaGKqVVCUBzG2SgGCPdyVzBqGJmutQP6-21x3Bw-qs6ZCoZYOB1J1pI1K1daAMAMr4M6pGvNSeXHIEZg4LRqA17IhGlxSmxFrlFAvsU7kZTeHRpd12pGaGNhdjHRYKdHIwBLqcxulAUjvtAbt1s8OtA4mTO+1xsHfWNTSOs3ZiF-8PNwsAw1lrRQSWO7h07kXbLfsA5mgKwMIrb7u6lZzgJvO1WPM93i5yFuuQ25NZKyZ8FoY0DIklAkAkYJ+OQuyMDBi9THr8cI8lTgAyE4sHjOJ0jVRagiGCOQdu11-yEsqknZdgJUncBQCpzDJnjMma4787IULSNiCwKRqzFB8EOZpFqpb6xnNzrO9qV7z6vNebUIdhIa2agbegmQYGLTgcegO3hgATAxYDp23OaaIFapTIAVPmHO1xy7V27kBrgHdkzD3ONceeyKJ9jmPsoC+3yLHv3rM0-exJjQh2Eeg+qOD+xW39Uc7h7bPDABmZHJ2zukYJ1dm7JPIVokJ4dhE9PvuuYMywP0UKsDvq82T9nsP1ci+5+tvnUP9VAA))

Reply:
@devongovett ping

Reply:
@devongovett  
If this topic requires additional investigation/discussion from the team, I can split PR in two, and this one will contain exports of TableCollection and TableKeyboardDelegate only

Reply:
Apologies about the delay, go ahead and split the PR into two for now.